chrome/content/zotero/xpcom/itemTreeManager.js

```diff
@@ -0,0 +1,370 @@
+		 * @param {(option: ItemTreeColumnOptions) => boolean} validator - A function that returns true if the option is valid
+		 * @returns {boolean} true if the option(s) are valid
+		 */
+		function validate(options, validator) {
+			if (Array.isArray(options)) {
+				return options.every(opt => validator(opt));
+			}
+			return validator(options);
+		}
+
+		// Check if the input option has duplicate dataKeys
+		const noInputDuplicates = !Array.isArray(options) || !options.find((opt, i, arr) => arr.findIndex(o => o.dataKey === opt.dataKey) !== i);
+		if (!noInputDuplicates) {
+			Zotero.warn(`ItemTree Column options have duplicate dataKey.`);
+		}
+		const requiredProperties = validate(options, (option) => {
+			const valid = option.dataKey && option.label;
+			if (!valid) {
+				Zotero.warn(`ItemTree Column option ${JSON.stringify(option)} must have dataKey and label.`);
+			}
+			return valid;
+		});
+		const noRegisteredDuplicates = validate(options, (option) => {
+			const valid = this.getColumns({ dataKey: option.dataKey }).length === 0;
+			if (!valid) {
+				Zotero.warn(`ItemTree Column option ${JSON.stringify(option)} with dataKey ${option.dataKey} already exists.`);
+			}
+			return valid;
+		});
+		return noInputDuplicates && requiredProperties && noRegisteredDuplicates;
+	}
+
+
+	/**
+	 * Add a new column or new columns.
+	 * If the options is an array, all its children must be valid.
+	 * Otherwise, no columns are added.
+	 * @param {ItemTreeColumnOptions | ItemTreeColumnOptions[]} options - An option or array of options to add
+	 * @returns {string | string[] | false} - The dataKey(s) of the added column(s) or false if no columns were added
+	 */
+	_addColumns(options) {
+		const isSingle = !Array.isArray(options);
+		if (isSingle) {
+			options = [options];
+		}
+		this._namespacedDataKey(options);
+		// If any check fails, return check results
+		if (!this._validateColumnOption(options)) {
+			return false;
+		}
+		for (const opt of options) {
+			this._customColumns[opt.dataKey] = Object.assign({}, opt, { custom: true });
+		}
+		return isSingle ? options[0].dataKey : options.map(opt => opt.dataKey);
+	}
+
+	/**
+	 * Remove a column option
+	 * @param {string | string[]} dataKeys 
+	 * @returns {boolean} - True if column(s) were removed, false if not
+	 */
+	_removeColumns(dataKeys) {
+		if (!Array.isArray(dataKeys)) {
+			dataKeys = [dataKeys];
+		}
+		// If any check fails, return check results and do not remove any columns
+		for (const key of dataKeys) {
+			if (!this._customColumns[key]) {
+				Zotero.warn(`ItemTree Column option with dataKey ${key} does not exist.`);
+				return false;
+			}
+		}
+		for (const key of dataKeys) {
+			delete this._customColumns[key];
+		}
+		return true;
+	}
+
+	/**
+	 * Make sure the dataKey is namespaced with the plugin ID
+	 * @param {ItemTreeColumnOptions | ItemTreeColumnOptions[]} options 
+	 * @returns {void}
+	 */
+	_namespacedDataKey(options) {
+		if (!Array.isArray(options)) {
```

Comment:
It's better to avoid functions that mutate objects because they are more difficult to reason about and test. This is why most built-in library functions (e.g. for strings, arrays) don't modify arguments but instead return new values.

Given that this function should either be named `namespaceDataKey()`, which would imply that it modifies the argument, or better yet, it should return the namespaced dataKey. It should take one argument and be used like
```
options.forEach(o => o.dataKey = this._namespacedDataKey(o));
```

Reply:
Done